compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala

```diff
@@ -950,9 +950,23 @@ class ClassfileParser(
           if (tastyBytes.nonEmpty) {
             val reader = new TastyReader(bytes, 0, 16)
             val expectedUUID = new UUID(reader.readUncompressedLong(), reader.readUncompressedLong())
-            val tastyUUID = new TastyHeaderUnpickler(tastyBytes).readHeader()
+            val tastyHeader = new TastyHeaderUnpickler(tastyBytes).readFullHeader()
+            val fileTastyVersion = TastyVersion(tastyHeader.majorVersion, tastyHeader.minorVersion, tastyHeader.experimentalVersion)
+            val tastyUUID = tastyHeader.uuid
             if (expectedUUID != tastyUUID)
               report.warning(s"$classfile is out of sync with its TASTy file. Loaded TASTy file. Try cleaning the project to fix this issue", NoSourcePosition)
+
+            val tastyFilePath = classfile.path.stripSuffix(".class") + ".tasty"
+            val isTastyCompatible =
+              TastyFormat.isVersionCompatible(fileVersion = fileTastyVersion, compilerVersion = ctx.tastyVersion) ||
+              classRoot.symbol.showFullName.startsWith("scala.") // References to stdlib are considered safe because we check the values of @since annotations
```

Comment:
I'm not sure what this escape hatch helps with, shouldn't `readFullHeader()` already throw if the version is not compatible?

Reply:
Even if only the scala stdlib used the scala package, this check still seems too broad: if the scala 3.2 compiler tries to read the stdlib of scala 3.3 and we actually changed tasty, it will crash when parsing the file.

Reply:
I was suspicious of the `@since` treatment before, and I think I now understand better *why*.

Clearly, as noted by @smarter, it is not sufficient to guarantee that an older compiler can read the TASTy of the file containing the new definition. In fact, it does not even guarantee that it could read the *other* stuff present in that file, even if it could jump over the new definitions entirely. It could very well be that the rest of the code in the file produces new TASTy as the result of different language rules, or because TASTy got a new feature that simplifies encoding, etc. Since the library is not compiled with `-scala-release`, its TASTy will not be tailored to be readable by an older compiler.

Reply:
Another, deeper problem with the `@since` treatment is that things like #14136 are fundamentally incompatible with it.

This wouldn't be an issue with the approach proposed by @julienrf at https://contributors.scala-lang.org/t/improving-scala-3-forward-compatibility/5298/4?u=sjrd (emit parts of the stdlib itself with `-scala-release`, and allow to depend on a newer stdlib than the compiler).

Reply:
@smarter maybe I missed something but in what situation would scala 3.2 compiler have to read TASTy of stdlib compiled with scala 3.3?

Reply:
If you add a dependency to a library compiled for 3.3 but you're compiling with 3.2 you should get an error while reading TASTy files of classes defined inside this library anyway so this shouldn't really matter that much, right?

Reply:
You might end up reading a tasty file from the standard library on the class path before reading one from the library you depend on.

Reply:
OK, so we could add an extra check, something like
```scala
val isTastyCompatible =
              TastyFormat.isVersionCompatible(fileTastyVersion, TastyVersion.compilerVersion) && (
                  TastyFormat.isVersionCompatible(fileTastyVersion, ctx.tastyVersion) ||
                  classRoot.symbol.showFullName.startsWith("scala.")
              )
```
or maybe handling these cases with different error messages. This should handle both the case when e.g.
a) we compile something with scala 3.3 compiler but with `-scala-release 3.2` and we want the compilation to fail when the code has some dependency to a library targeting 3.3
b) we use scala 3.2 compiler to compile code  which for whatever reason brings scala 3.3 stdlib to the classpath

Reply:
I gave up the idea of more precise checks for prefixes. This seems not to be very flexible as some definitions are added directly to `scala` package so having to keep track of that is probably not worth the effort